### PR TITLE
Add the retrive function to fetch single files

### LINKED ISSUE #147: Create a function to fetch a single file
**Description of the desired feature**

Consider the following use case: 

> I'm a scientist/educator writing a notebook and I need to get data from this one file on the internet. I'm on Jupyter Lab, possibly on a remote server. I don't know the hash of the file and don't need to download many files, just this one. I want the file to be cached locally and not downloaded multiple times. After the first download, I want to set the hash so that I can make sure future runs of the notebook use the same version of the data. 

Right now, if you want to use Pooch for that you need to write this code:

```python
from unittest import mock

url = "https://www.ngdc.noaa.gov/mgg/ocean_age/data/2008/grids/age/age.3.6.nc.bz2"
remote_file = url.split("/")[-1]

fname = pooch.create(
    path=pooch.os_cache("meh"), 
    base_url="", 
    registry={remote_file: mock.ANY}, 
    urls={remote_file: url}
).fetch(remote_file)

print(pooch.file_hash(fname))
# Copy and paste the hash in the registry above for future runs.
```
There is a lot of repetition and feels like a hack. The same thing in fsspec would look like:

```python
# files is a not-yet-open OpenFile object. The "with" context actually opens it
with fsspec.open(url, mode="r") as f:
    # Read in the data
```

which is much more readable but it doesn't seem to cache automatically (requiring further setup) and doesn't have a hash checking mechanism (at least not that I could find in the docs).

**Proposal**

Add a function `pooch.retrieve` that would make the code above:

```python
fname = pooch.retrieve(url, hash=None)
# hash of downloaded file logged to screen and copy+pasted above for future runs
```

* Automatically cache to a `pooch.os_cache("pooch")` location without further setup.
* File names would be set to a short hash of the file content to make sure no overwriting happens. 
* The return is the file name only, keeping with the pooch way of not trying to do anything fancy. 
* Automatically printing out the hash makes it easier to download first then make sure future downloads are the same.
* Cache location and file name can be set by user if desired.
* Function takes `downloader` and `processor` arguments like `pooch.Pooch.fetch`.

This makes it easy to have all the features of Pooch for single file downloads in very little code.

The name `pooch.fetch` is nicer but it may cause confusion with `pooch.Pooch.fetch` so `retrieve` is a good alternative.

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by leouieda**
Add the function `retrieve` for downloading single data files. This is much more convenient than setting up a `Pooch`. It automatically selects a unique file name and saves files to a "pooch" cache (as determined by `os_cache`). Allows passing in `known_hash=None` to bypass the hash checks. In this case, will print a message with the SHA256 hash of the downloaded file and advising users to set `known_hash` to this value.

Fixes #147 

**Reminders**:

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst` and the base `__init__.py` file for the package.
- [ ] Write detailed docstrings for all functions/classes/methods. It often helps to design better code if you write the docstrings first.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.
- [ ] Add your full name, affiliation, and [ORCID](https://orcid.org) (optional) to the `AUTHORS.md` file (if you haven't already) in case you'd like to be listed as an author on the [Zenodo](https://zenodo.org/communities/fatiando) archive of the next release.

### CODE REVIEW

**Comment by leouieda** [comment]
TODO:

- [ ] Name files with `hash-filename.ext` to make sure the extension is preserved and the cache is human readable.

**Comment by leouieda** [comment]
This got fairly big so I'll separate it into smaller PRs to make them easier to review.

**Comment by leouieda** [comment]
Alright, this is pretty much done and tests pass everywhere. I'd appreciate a review if anyone has the time.

**Comment by leouieda** [comment]
> Wouldn't be nice to have an optional argument to download the file even if a version of it already exists in the cache? This could be useful for files that are regularly updated. What do you think?

Do you mean like a `force=True` argument to download the file regardless? I can see the point but I wonder if this wouldn't make it too close to what `urllib` already does. The main feature here is the hash checking and caching so I'm hesitant to include something that shouldn't be used too often.

What do you think?

File: pooch/core.py [hunk 1]
```diff
@@ -15,10 +15,221 @@
...
+        prepend the hash with ``algorithm:``, for example
+        ``md5:pw9co2iun29juoh`` or ``sha1:092odwhi2ujdp2du2od2odh2wod2``. If
+        None, will NOT check the hash of the downloaded file or check if an
+        existing file needs to be updated.
+    fname : str or None
+        The name that will be used to save the file. Should NOT include the
+        full the path, just the file name (it will be appended to *path*). If
+        None, will create a unique file name using a combination of the last
+        part of the URL (assuming it's the file name) and the MD5 hash of the
+        URL. For example, ``81whdo2d2e928yd1wi22:data-file.csv``. This ensures
```
[the remark this answers is not in the document]
Comment (leouieda) [reply pooch/core.py]: Nice catch! Thanks

**Comment by santisoler** [comment]
> Do you mean like a `force=True` argument to download the file regardless? I can see the point but I wonder if this wouldn't make it too close to what `urllib` already does. The main feature here is the hash checking and caching so I'm hesitant to include something that shouldn't be used too often.
> 
> What do you think?

I was thinking that would be nice to have such option in cases where we still want to use decompressors, or when we want to redownload the file without manually removing the cached file.
Yet, I don't know how often this will be used.

If you want to take time to think about this, we can merge this and add this later.

**Comment by leouieda** [comment]
> If you want to take time to think about this, we can merge this and add the new argument later.

Right, you have a good point. I think I prefer to release this first and wait to see if the need arises. It's easier to add an option than to take it away later :)

File: README.rst [hunk 2]
```diff
@@ -68,6 +68,25 @@ Here are Pooch's main features:
 Example
 -------
 
+For a **scientist downloading a data file** for analysis:
```
Comment (leouieda) [inline README.rst:71]:
```suggestion
A **scientist downloading a data file** for analysis:
```

File: README.rst [hunk 3]
```diff
@@ -68,6 +68,25 @@ Here are Pooch's main features:
...
+    # a download. Pooch will check the hash (checksum) of the downloaded file
+    # against the given value to make sure it's the right file (not corrupted
+    # or outdated).
+    fname = retrieve(
+        url="https://some-data-server.org/a-data-file.nc",
+        known_hash="md5:70e2afd3fd7e336ae478b1e740a5f08e",
+    )
+
+
+For **package developers** including sample data in their projects:
```
Comment (leouieda) [inline README.rst:88]:
```suggestion
A **package developer** including sample data in their project:
```

**Pull request merged**